cornell-dti/course-plan
# Say goodbye to dragula and fix a bunch of drag and drop issues

### Summary 

`vue2-dragula` is unmaintained. It's too opinionated so that we can't get a lot of things we want. It also doesn't provide good type definitions. On the other hand [Vue.Draggable](https://github.com/SortableJS/Vue.Draggable) is much better maintained, and it allows us to customize the copy function so we can allocate unique ID on demand.

This PR replaces `vue2-dragula` with `Vue.Draggable`. Surprising, quite a few issues can now be fixed easily.

1. Double course card found in #276. Fixed automatically, not sure why but maybe `vue2-dragula` is just bad.
2. We can now ban drag and drop between requirement bar courses by using `:value` instead of `v-model`. This ensures that the courses array are only read and not written, so all the drop effects are cancelled automatically.
3. We no longer need to watch the changes for the `courses` array in `Semester.vue`. We can use a settable computed property, where the set function emits a message asking parent for update instead of updating itself. This removes the last place where the semester array is updated implicitly.

With this setup in place, we can do the followings in future PRs:

1. Allocate uniqueID of courses only after drop
2. Integrate semesters into global data listeners

### Test Plan 

- drag and drop within and between semesters
- drag and drop from incomplete courses and see all
- reload to verify all data is auto saved

### Notes

Vue.Draggable also has plans to support vue 3: https://github.com/SortableJS/vue.draggable.next
This is nice because maybe we don't want to stuck on vue 2 forever.
